File: wasm-extension-js/src/exports.ts
// Moosync
// Copyright (C) 2024, 2025  Moosync <support@moosync.app>
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

import { callListener } from "./api";
import { ExtensionEventName } from './types';

/**
 * Must be spread within the Extism `module.exports`. Your bundler/build process should handle this for you.
 */
export namespace Exports {
  async function handleWrapper(event: ExtensionEventName) {
    const input = Host.inputString();
    let params = undefined;
    if (input) {
      params = JSON.parse(input);
    }

    try {
      let resp = await callListener(event, params);
      if (typeof resp !== "undefined") {
        Host.outputString(JSON.stringify(resp));
      } else {
        Host.outputString(JSON.stringify(null));
      }
    } catch (err) {
      console.error("caught err", err);
      Host.outputString(JSON.stringify(null));
      return;
    }
  }

  export function get_provider_scopes_wrapper() {
    // Internal function
    handleWrapper("getProviderScopes" as ExtensionEventName);
  }
  export function get_playlists_wrapper() {
    handleWrapper("getPlaylists");
  }
  export function get_playlist_content_wrapper() {
    handleWrapper("getPlaylistContent");
  }
  export function get_playlist_from_url_wrapper() {
    handleWrapper("getPlaylistFromUrl");
  }
  export function get_playback_details_wrapper() {
    handleWrapper("getPlaybackDetails");
  }
  // TODO: rename this to get_search_wrapper once the app is updated
  export function search_wrapper() {
    // Legacy name
    handleWrapper("search" as ExtensionEventName);

    handleWrapper("getSearch");
  }
  export function get_recommendations_wrapper() {
    handleWrapper("getRecommendations");
  }
  export function get_song_from_url_wrapper() {
    handleWrapper("getSongFromUrl");
  }
  // TODO: rename this to get_stream_url_wrapper once the app is updated
  export function handle_custom_request_wrapper() {
    // Legacy name
    handleWrapper("handleCustomRequest" as ExtensionEventName);

    handleWrapper("getStreamUrl");
  }
  export function get_artist_songs_wrapper() {
    handleWrapper("getArtistSongs");
  }
  export function get_album_songs_wrapper() {
    handleWrapper("getAlbumSongs");
  }
  export function get_song_from_id_wrapper() {
    handleWrapper("getSongFromId");
  }
  export function on_queue_changed_wrapper() {
    handleWrapper("onQueueChanged");
  }
  export function on_volume_changed_wrapper() {
    handleWrapper("onVolumeChanged");
  }
  export function on_player_state_changed_wrapper() {
    handleWrapper("onPlayerStateChanged");
  }
  export function on_song_changed_wrapper() {
    handleWrapper("onSongChanged");
  }
  export function on_seeked_wrapper() {
    handleWrapper("onSeeked");
  }
  export function on_preferences_changed_wrapper() {
    handleWrapper("onPreferencesChanged");
  }
  export function on_song_added_wrapper() {
    handleWrapper("onSongAdded");
  }
  export function on_song_removed_wrapper() {
    handleWrapper("onSongRemoved");
  }
  export function on_playlist_added_wrapper() {
    handleWrapper("onPlaylistAdded");
  }
  export function on_playlist_removed_wrapper() {
    handleWrapper("onPlaylistRemoved");
  }
  export function get_accounts_wrapper() {
    handleWrapper("getAccounts");
  }
  export function perform_account_login_wrapper() {
    handleWrapper("performAccountLogin");
  }
  // TODO: rename this to on_scrobble_wrapper once the app is updated
  export function scrobble_wrapper() {
    // Legacy name
    handleWrapper("scrobble" as ExtensionEventName);

    handleWrapper("onScrobble");
  }
  // TODO: rename this to on_oauth_success_wrapper once the app is updated
  export function oauth_callback_wrapper() {
    // Legacy name
    handleWrapper("oauthCallback" as ExtensionEventName);

    handleWrapper("onOauthSuccess");
  }
  export function get_song_context_menu_wrapper() {
    handleWrapper("getSongContextMenu");
  }
  export function get_playlist_context_menu_wrapper() {
    handleWrapper("getPlaylistContextMenu");
  }
  export function on_context_menu_action_wrapper() {
    handleWrapper("onContextMenuAction");
  }
  export function get_lyrics_wrapper() {
    handleWrapper("getLyrics");
  }
}

File: wasm-extension-go/pkg/api/exports.go
package api

import (
	"encoding/json"

	"github.com/Moosync/extensions-sdk/wasm-extension-go/pkg/types"
	"github.com/extism/go-pdk"
)

var extension Extension

func get_params(input any) int32 {
	if err := pdk.InputJSON(input); err != nil {
		pdk.SetError(err)
		return 1
	}
	return 0
}

func runWrapper(input any, callback func() (any, error)) int32 {
	if input != nil {
		if get_params(input) != 0 {
			return 1
		}
	}
	result, err := callback()
	if err != nil {
		pdk.SetError(err)
		return 1
	}
	if result != nil {
		jsonString, err := json.Marshal(result)
		LogInfo("error %v", err)
		if err != nil {
			pdk.SetError(err)
			return 1
		}
		pdk.Output(jsonString)
	}
	return 0
}

//go:wasmexport get_provider_scopes_wrapper
func get_provider_scopes_wrapper() int32 {
	return runWrapper(nil, func() (any, error) {
		return extension.GetProviderScopes()
	})
}

//go:wasmexport get_playlists_wrapper
func get_playlists_wrapper() int32 {
	return runWrapper(nil, func() (any, error) {
		playlists, err := extension.GetPlaylists()
		if err != nil {
			return nil, err
		}
		return types.PlaylistReturnType{Playlists: playlists}, nil
	})
}

//go:wasmexport get_playlist_content_wrapper
func get_playlist_content_wrapper() int32 {
	var id string
	var token string
	in := [...]any{&id, &token}
	return runWrapper(&in, func() (any, error) {
		songs, err := extension.GetPlaylistContent(id, token)
		if err != nil {
			return nil, err
		}
		return types.SongsWithPageTokenReturnType{
			Songs:         songs,
			NextPageToken: nil,
		}, nil
	})
}

//go:wasmexport get_playlist_from_url_wrapper
func get_playlist_from_url_wrapper() int32 {
	var url string
	in := [...]any{&url}
	return runWrapper(&in, func() (any, error) {
		playlist, err := extension.GetPlaylistFromURL(url)
		if err != nil {
			return nil, err
		}
		return types.PlaylistAndSongsReturnType{
			Playlist: &playlist,
			Songs:    nil,
		}, nil
	})
}

//go:wasmexport get_playback_details_wrapper
func get_playback_details_wrapper() int32 {
	var song types.Song
	in := [...]any{&song}
	return runWrapper(&in, func() (any, error) {
		return extension.GetPlaybackDetails(song)
	})
}

//go:wasmexport search_wrapper
func search_wrapper() int32 {
	var term string
	return runWrapper(&term, func() (any, error) {
		result, err := extension.Search(term)
		if err != nil {
			return nil, err
		}

		songs := result.Songs
		if songs == nil {
			songs = []types.Song{}
		}

		playlists := result.Playlists
		if playlists == nil {
			playlists = []types.QueryablePlaylist{}
		}

		artists := result.Artists
		if artists == nil {
			artists = []types.QueryableArtist{}
		}

		albums := result.Albums
		if albums == nil {
			albums = []types.QueryableAlbum{}
		}

		// genres := result.Genres
		// if genres == nil {
		// 	genres = []types.QueryableGenre{}
		// }
		return types.SearchReturnType{
			Songs:     songs,
			Playlists: playlists,
			Artists:   artists,
			Albums:    albums,
		}, nil
	})
}

//go:wasmexport get_recommendations_wrapper
func get_recommendations_wrapper() int32 {
	return runWrapper(nil, func() (any, error) {
		songs, err := extension.GetRecommendations()
		if err != nil {
			return nil, err
		}
		return types.RecommendationsReturnType{Songs: songs}, nil
	})
}

//go:wasmexport get_song_from_url_wrapper
func get_song_from_url_wrapper() int32 {
	var url string
	in := [...]any{&url}
	return runWrapper(&in, func() (any, error) {
		song, err := extension.GetSongFromURL(url)
		if err != nil {
			return nil, err
		}
		return types.SongReturnType{Song: &song}, nil
	})
}

//go:wasmexport handle_custom_request_wrapper
func handle_custom_request_wrapper() int32 {
	var url string
	return runWrapper(&url, func() (any, error) {
		res, err := extension.HandleCustomRequest(url)
		if err != nil {
			return nil, err
		}
		return res, nil
	})
}

//go:wasmexport get_artist_songs_wrapper
func get_artist_songs_wrapper() int32 {
	var artist types.QueryableArtist
	var token string
	in := [...]any{&artist, &token}
	return runWrapper(&in, func() (any, error) {
		songs, err := extension.GetArtistSongs(artist, token)
		if err != nil {
			return nil, err
		}
		return types.SongsWithPageTokenReturnType{
			Songs:         songs,
			NextPageToken: nil,
		}, nil
	})
}

//go:wasmexport get_album_songs_wrapper
func get_album_songs_wrapper() int32 {
	var album types.QueryableAlbum
	var token string
	in := [...]any{&album, &token}
	return runWrapper(&in, func() (any, error) {
		songs, err := extension.GetAlbumSongs(album, token)
		if err != nil {
			return nil, err
		}
		return types.SongsWithPageTokenReturnType{
			Songs:         songs,
			NextPageToken: nil,
		}, nil
	})
}

//go:wasmexport get_song_from_id_wrapper
func get_song_from_id_wrapper() int32 {
	var id string
	return runWrapper(&id, func() (any, error) {
		song, err := extension.GetSongFromID(id)
		if err != nil {
			return nil, err
		}
		return types.SongReturnType{Song: &song}, nil
	})
}

//go:wasmexport on_queue_changed_wrapper
func on_queue_changed_wrapper() int32 {
	var queue types.Value
	in := [...]any{&queue}
	return runWrapper(&in, func() (any, error) {
		return nil, extension.OnQueueChanged(queue)
	})
}

//go:wasmexport on_volume_changed_wrapper
func on_volume_changed_wrapper() int32 {
	return runWrapper(nil, func() (any, error) {
		return nil, extension.OnVolumeChanged()
	})
}

//go:wasmexport on_player_state_changed_wrapper
func on_player_state_changed_wrapper() int32 {
	return runWrapper(nil, func() (any, error) {
		return nil, extension.OnPlayerStateChanged()
	})
}

//go:wasmexport on_song_changed_wrapper
func on_song_changed_wrapper() int32 {
	return runWrapper(nil, func() (any, error) {
		return nil, extension.OnSongChanged()
	})
}

//go:wasmexport on_seeked_wrapper
func on_seeked_wrapper() int32 {
	var t float64
	in := [...]any{&t}
	return runWrapper(&in, func() (any, error) {
		return nil, extension.OnSeeked(t)
	})
}

//go:wasmexport on_preferences_changed_wrapper
func on_preferences_changed_wrapper() int32 {
	var args types.PreferenceArgs
	in := [...]any{&args}
	return runWrapper(&in, func() (any, error) {
		return nil, extension.OnPreferencesChanged(args)
	})
}

//go:wasmexport on_song_added_wrapper
func on_song_added_wrapper() int32 {
	var song types.Song
	in := [...]any{&song}
	return runWrapper(&in, func() (any, error) {
		return nil, extension.OnSongAdded(song)
	})
}

//go:wasmexport on_song_removed_wrapper
func on_song_removed_wrapper() int32 {
	var song types.Song
	in := [...]any{&song}
	return runWrapper(&in, func() (any, error) {
		return nil, extension.OnSongRemoved(song)
	})
}

//go:wasmexport on_playlist_added_wrapper
func on_playlist_added_wrapper() int32 {
	var playlist types.QueryablePlaylist
	in := [...]any{&playlist}
	return runWrapper(&in, func() (any, error) {
		return nil, extension.OnPlaylistAdded(playlist)
	})
}

//go:wasmexport on_playlist_removed_wrapper
func on_playlist_removed_wrapper() int32 {
	var playlist types.QueryablePlaylist
	in := [...]any{&playlist}
	return runWrapper(&in, func() (any, error) {
		return nil, extension.OnPlaylistRemoved(playlist)
	})
}

//go:wasmexport get_accounts_wrapper
func get_accounts_wrapper() int32 {
	return runWrapper(nil, func() (any, error) {
		accounts, err := extension.GetAccounts()
		if err != nil {
			return nil, err
		}
		return accounts, nil
	})
}

//go:wasmexport perform_account_login_wrapper
func perform_account_login_wrapper() int32 {
	var args types.AccountLoginArgs
	in := [...]any{&args}
	return runWrapper(&in, func() (any, error) {
		res, err := extension.PerformAccountLogin(args)
		if err != nil {
			return nil, err
		}
		return res, nil
	})
}

//go:wasmexport scrobble_wrapper
func scrobble_wrapper() int32 {
	var song types.Song
	in := [...]any{&song}
	return runWrapper(&in, func() (any, error) {
		return nil, extension.Scrobble(song)
	})
}

//go:wasmexport oauth_callback_wrapper
func oauth_callback_wrapper() int32 {
	var code string
	return runWrapper(&code, func() (any, error) {
		return nil, extension.OauthCallback(code)
	})
}

//go:wasmexport get_song_context_menu_wrapper
func get_song_context_menu_wrapper() int32 {
	var songs []types.Song
	in := [...]any{&songs}
	return runWrapper(&in, func() (any, error) {
		menu, err := extension.GetSongContextMenu(songs)
		if err != nil {
			return nil, err
		}
		return menu, nil
	})
}

//go:wasmexport get_playlist_context_menu_wrapper
func get_playlist_context_menu_wrapper() int32 {
	var playlist types.QueryablePlaylist
	in := [...]any{&playlist}
	return runWrapper(&in, func() (any, error) {
		menu, err := extension.GetPlaylistContextMenu(playlist)
		if err != nil {
			return nil, err
		}
		return menu, nil
	})
}

//go:wasmexport on_context_menu_action_wrapper
func on_context_menu_action_wrapper() int32 {
	var action string
	return runWrapper(&action, func() (any, error) {
		return nil, extension.OnContextMenuAction(action)
	})
}

//go:wasmexport get_lyrics_wrapper
func get_lyrics_wrapper() int32 {
	var song types.Song
	in := [...]any{&song}
	return runWrapper(&in, func() (any, error) {
		lyrics, err := extension.GetLyrics(song)
		if err != nil {
			return nil, err
		}
		return lyrics, nil
	})
}

//go:wasmimport extism:host/user send_main_command
func send_main_command(uint64) uint64

//go:wasmimport extism:host/user system_time
func system_time() uint64

//go:wasmimport extism:host/user open_clientfd
func open_clientfd(uint64) uint64

//go:wasmimport extism:host/user write_sock
func write_sock(int64, uint64) uint64

//go:wasmimport extism:host/user read_sock
func read_sock(int64, uint64) uint64

//go:wasmimport extism:host/user hash
func hash(uint64, uint64) uint64
